services/venice-controller/src/main/java/com/linkedin/venice/controller/systemstore/SystemStoreHealthCheckService.java

```diff
@@ -0,0 +1,164 @@
+import com.linkedin.venice.exceptions.VeniceException;
+import com.linkedin.venice.meta.ReadWriteStoreRepository;
+import com.linkedin.venice.meta.Store;
+import com.linkedin.venice.pushstatushelper.PushStatusStoreReader;
+import com.linkedin.venice.pushstatushelper.PushStatusStoreWriter;
+import com.linkedin.venice.service.AbstractVeniceService;
+import com.linkedin.venice.system.store.MetaStoreReader;
+import com.linkedin.venice.system.store.MetaStoreWriter;
+import com.linkedin.venice.utils.RetryUtils;
+import com.linkedin.venice.utils.concurrent.VeniceConcurrentHashMap;
+import java.time.Duration;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+
+/**
+ * This class is used to schedule periodic check of user system stores in Venice cluster to make sure system stores has
+ * good version and is ingesting correctly. This service is expected to be setup in the leader controller of the cluster,
+ * and it will periodically scan all user system stores and collect bad system stores. Parent controller will hit the
+ * controller endpoint to collect information about bad system stores.
+ */
+public class SystemStoreHealthCheckService extends AbstractVeniceService {
+  public static final Logger LOGGER = LogManager.getLogger(SystemStoreHealthCheckService.class);
+  private final ReadWriteStoreRepository storeRepository;
+  private final PushStatusStoreReader pushStatusStoreReader;
+  private final PushStatusStoreWriter pushStatusStoreWriter;
+  private final MetaStoreReader metaStoreReader;
+  private final MetaStoreWriter metaStoreWriter;
+
+  private final int checkPeriodInSeconds;
+  private final AtomicBoolean isRunning = new AtomicBoolean(false);
+  private ScheduledExecutorService checkServiceExecutor;
+
+  private Set<String> unhealthySystemStoreSet = new HashSet<>();
+
+  public SystemStoreHealthCheckService(
+      ReadWriteStoreRepository storeRepository,
+      MetaStoreReader metaStoreReader,
+      MetaStoreWriter metaStoreWriter,
+      PushStatusStoreReader pushStatusStoreReader,
+      PushStatusStoreWriter pushStatusStoreWriter,
+      int systemStoreCheckPeriodInSeconds) {
+    this.storeRepository = storeRepository;
+    this.metaStoreWriter = metaStoreWriter;
+    this.metaStoreReader = metaStoreReader;
+    this.pushStatusStoreWriter = pushStatusStoreWriter;
+    this.pushStatusStoreReader = pushStatusStoreReader;
+    this.checkPeriodInSeconds = systemStoreCheckPeriodInSeconds;
+  }
+
+  /**
+   * Return unhealthy system store name set. This API is expected to be called by parent controller.
+   */
+  public Set<String> getUnhealthySystemStoreSet() {
```

Comment:
I am wondering if strict visibility is needed here:
The getter will be used in the followup PR for getting the unhealthy system stores from parent controller call. This will be periodic (1h maybe) and this child periodic job will also be something like once per hour. The set will be updated, but it is OK so see the old copy. Are you saying that without volatile, it may never see the update in memory from another thread?

Reply:
My understanding is that without volatile/AtomicReference, there is no guarantee in when the updates (cause we are doing reference re-assignment periodically) can be seen by another thread. The thread is free to cache this value locally and I probably won't be surprised if it cached it for a very long time.

Reply:
Sure, I see your concern, I can update that.